octopusdeploy_framework/resource_tenant.go

```diff
@@ -0,0 +1,170 @@
+}
+
+func (r *tenantTypeResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data *schemas.TenantModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tflog.Info(ctx, fmt.Sprintf("reading Tenant (%s)", data.ID))
+
+	client := r.Config.Client
+	tenant, err := tenants.GetByID(client, data.SpaceID.ValueString(), data.ID.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("unable to load tenant", err.Error())
+		return
+	}
+
+	mapTenantToState(data, tenant)
+
+	tflog.Info(ctx, fmt.Sprintf("Tenant read (%s)", tenant.GetID()))
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *tenantTypeResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	internal.Mutex.Lock()
+	defer internal.Mutex.Unlock()
+
+	var data, state *schemas.TenantModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tflog.Debug(ctx, fmt.Sprintf("updating tenant '%s'", data.ID.ValueString()))
+
+	tenantFromApi, err := tenants.GetByID(r.Config.Client, data.SpaceID.ValueString(), data.ID.ValueString())
+
+	tenant, err := mapStateToTenant(data)
+	tenant.ID = state.ID.ValueString()
+	if err != nil {
+		resp.Diagnostics.AddError("unable to load tenant", err.Error())
+		return
+	}
+
+	tflog.Info(ctx, fmt.Sprintf("updating Tenant (%s)", data.ID))
+
+	tenant.ProjectEnvironments = tenantFromApi.ProjectEnvironments
+	updatedTenant, err := tenants.Update(r.Config.Client, tenant)
+	if err != nil {
+		resp.Diagnostics.AddError("unable to update tenant", err.Error())
+		return
+	}
+
+	mapTenantToState(data, updatedTenant)
+
+	tflog.Info(ctx, fmt.Sprintf("Tenant updated (%s)", data.ID))
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *tenantTypeResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	internal.Mutex.Lock()
+	defer internal.Mutex.Unlock()
+
+	var data schemas.TenantModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if err := tenants.DeleteByID(r.Config.Client, data.SpaceID.ValueString(), data.ID.ValueString()); err != nil {
+		resp.Diagnostics.AddError("unable to delete tenant", err.Error())
+		return
+	}
+}
+
+func mapStateToTenant(data *schemas.TenantModel) (*tenants.Tenant, error) {
+	tenant := tenants.NewTenant(data.Name.ValueString())
+	tenant.ID = data.ID.ValueString()
+	tenant.ClonedFromTenantID = data.ClonedFromTenantId.ValueString()
+	tenant.Description = data.Description.ValueString()
+	tenant.SpaceID = data.SpaceID.ValueString()
+	if len(data.TenantTags.Elements()) > 0 {
+		tenant.TenantTags = util.ExpandStringList(data.TenantTags)
+	} else {
+		tenant.TenantTags = []string{}
+	}
+	sort.Strings(tenant.TenantTags)
```

Comment:
These sorts are a little odd, but I was receiving inconsistent input/output and this seems like the only viable fix.
![image](https://github.com/user-attachments/assets/926ae949-59f5-4777-b81c-83213b5ee05a)